plugins/BEdita/Core/tests/TestCase/Model/Table/TreesTableTest.php

```diff
@@ -56,29 +82,134 @@ public function tearDown()
+                true,
+                null,
+            ],
+            'folder' => [
+                true,
+                12,
+            ],
+            'notAFolder' => [
+                false,
+                4,
+            ],
+        ];
     }
 
     /**
-     * Test validationDefault method
+     * Test for `isParentValid()`
      *
      * @return void
+     *
+     * @dataProvider isParentValidProvider
+     * @covers ::isParentValid()
      */
-    public function testValidationDefault()
+    public function testIsParentValid($expected, $parentId)
     {
-        $this->markTestIncomplete('Not implemented yet.');
+        $entity = $this->Trees->newEntity();
+        $entity->parent_id = $parentId;
+        static::assertEquals($expected, $this->Trees->isParentValid($entity));
     }
 
     /**
-     * Test buildRules method
+     * Data provider for `testChangeRoot()`
      *
      * @return void
      */
-    public function testBuildRules()
+    public function changeRootProvider()
     {
-        $this->markTestIncomplete('Not implemented yet.');
+        return [
+            'becomeRoot' => [
+                12,
+                null,
+            ],
+            'changeRoot' => [
+                13,
+                13,
+            ],
+        ];
+    }
+
+    /**
+     * Test that moving a node under another `root_id`
+     * all children will be migrated to the same `root_id`
+     *
+     * @return void
+     *
+     * @dataProvider changeRootProvider
+     * @covers ::afterSave()
+     */
+    public function testChangeRoot($rootExpected, $parentId)
+    {
+        $node = $this->Trees->get(2);
+        static::assertEquals(11, $node->root_id);
+        $children = $this->Trees->find('children', ['for' => 2])->toList();
+
+        $node->parent_id = $parentId;
+        static::assertTrue((bool)$this->Trees->save($node));
+
+        $node = $this->Trees->get(2);
+        $actualChildren = $this->Trees->find('children', ['for' => 2])->toList();
+
+        static::assertEquals($rootExpected, $node->root_id);
+        static::assertCount(count($children), $actualChildren);
+        foreach ($actualChildren as $child) {
+            static::assertEquals($rootExpected, $child->root_id);
+        }
+    }
+
+    /**
+     * Test that moving a parent as child fails.
+     *
+     * @return void
+     * @coversNothing
+     * @expectedException \RuntimeException
+     */
+    public function testMoveParentAsChild()
+    {
+        // create new Folder
+        LoggedUser::setUser(['id' => 1]);
+        $Folders = TableRegistry::get('Folders');
+        $entity = $Folders->newEntity(['title' => 'subsub folder']);
+        $entity->type = 'folders';
+        $entity->parent = $Folders->get(12);
+
+        $Folders->save($entity);
+
+        $parentNode = $this->Trees
+            ->find()
+            ->where(['object_id' => $entity->parent->id])
+            ->first();
+
+        $node = $this->Trees
```

Comment:
This value is not used anywhere… maybe an assertion is missing?

Reply:
Naaaa... not needed, I remove it